feat(plugins): Adding colors to BigNumber with Time Comparison chart

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Currently our BigNumber with Time Comparison chart has no visual indicators to help user easily tell when the differences are positive or negatives. This PR adds a customize control to enable colors in the chart.

Also this PR is updating the thumbnail to reflect an up to date version of the chart with these colors.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1396" alt="Screenshot 2024-02-08 at 15 21 08" src="https://github.com/apache/superset/assets/38889534/9a4b7f32-1741-4b4a-ada2-7da2d658870d">


### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Add new checbox control so users can toggle colors in the chart
- Make the component capable of using the right colors based on the data
- Fix types in props ofthe component
- Update the thumbnail of the chart

kgabryje commented on superset-frontend/plugins/plugin-chart-period-over-period-kpi/src/PopKPI.tsx:
```diff
@@ -30,16 +34,51 @@ const ComparisonValue = styled.div<PopKPIComparisonValueStyleProps>`
+        comparisonColorEnabled ? backgroundColor : defaultBackgroundColor
+      };
+      color: ${comparisonColorEnabled ? textColor : defaultTextColor};
+      padding: 2px 5px;
```
could we use full grid units? Or does it look weird?

kgabryje commented on superset-frontend/plugins/plugin-chart-period-over-period-kpi/src/PopKPI.tsx:
```diff
@@ -30,16 +34,51 @@ const ComparisonValue = styled.div<PopKPIComparisonValueStyleProps>`
 
+const SymbolWrapper = styled.div<PopKPIComparisonSymbolStyleProps>`
+  ${({ theme, percentDifferenceNumber, comparisonColorEnabled }) => {
+    const defaultBackgroundColor = theme.colors.grayscale.light4;
```
Interesting approach with putting js inside of styled component like that! But I wonder if deducing colours from more "high-level" data belongs to a styled component. Maybe we should move that logic to `PopKPI` and just pass `backgroundColor` and `textColor` to `SymbolWrapper`?
But if you think this solution is better then I won't argue, I don't have a strong opinion here

Antonio-RiveroMartnez replied:
hehe yeah, I had it like that before and moved to this, it doesn't feel quite "well" but now that you noticed too, let's move it there instead. Thanks

rusackas commented:
I'm trying to upgrade Storybook, but ironing out a few last details. When that's done, we should add this plugin to Storybook and make sure it has controls for these options

kgabryje commented on superset-frontend/plugins/plugin-chart-period-over-period-kpi/src/PopKPI.tsx:
```diff
@@ -77,18 +135,26 @@ export default function PopKPI(props: PopKPIProps) {
+                ? prevNumber
+                : index === 1
+                ? valueDifference
+                : percentDifferenceFormattedString}
```
Hmm I'm not sure about coupling the values with indexes like this. I think it would be cleaner if we had some key-value mapping between symbols and values, or an array like [['#', prevNumber], ['△', valueDifference], ['%', percentDifferenceFormattedString]]

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Move background and textcolor computation inside the main component so the Styled doesnt have to deal with logic
- Stop accessing values with indexes and have a single array that collects both the symbol and the value
- fix more typing

yousoph commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

yousoph commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

mistercrunch commented:
Hey, quick note that I'm trying to fix ephemeral envs here -> https://github.com/apache/superset/pull/27057, hoping to merge quickly as soon as CI passes, and to confirm the fix on this PR right after...

mistercrunch commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

mistercrunch commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

mistercrunch commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

mistercrunch commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Update Thumbnail to be square so it doest look squished

Antonio-RiveroMartnez commented:
/testenv up FEATURE_CHART_PLUGINS_EXPERIMENTAL=true

kgabryje commented:
Nit for eventual follow up PR - add advanced analytics tag? cc @yousoph 

<img width="489" alt="image" src="https://github.com/apache/superset/assets/15073128/0123b57b-4f64-4e56-b92c-31fa92db5d5b">

kgabryje commented:
I think the selected number formatting doesn't apply to %

<img width="963" alt="image" src="https://github.com/apache/superset/assets/15073128/4f67cc87-2477-40f9-ac48-6012c7e1de0f">

kgabryje commented:
The styling is not responsive - when I added the chart to a dashboard in a small container, I can't see the values on the right (the container is also not scrollable horizontally so they're just hidden)

<img width="634" alt="image" src="https://github.com/apache/superset/assets/15073128/7a933924-de4b-4754-a0b2-dccd2f01812d">

kgabryje commented:
1. Can we add a description tooltip?
2. Does week/month/year mean last week/last month/last year? If so, maybe we could add "last" prefix to the names?

<img width="322" alt="image" src="https://github.com/apache/superset/assets/15073128/8c5d124f-ba2d-469d-a9a0-3d33ee184b0d">

kgabryje commented:
1. Since "Filters (custom)" control works only if Range For Comparison is "Custom", can we make the visibility of that control dynamic, so that it's actually only visible when it's useful? I believe we use such pattern in some other charts.
2. Can we remove the collapsible and maybe change the name of the control to sth like "Custom comparison range"?

<img width="327" alt="image" src="https://github.com/apache/superset/assets/15073128/ce13795d-8095-4598-ad94-55ae06a8a177">

Antonio-RiveroMartnez commented:
> 1. Can we add a description tooltip?
> 2. Does week/month/year mean last week/last month/last year? If so, maybe we could add "last" prefix to the names?
> 
> <img alt="image" width="322" src="https://private-user-images.githubusercontent.com/15073128/304134536-8c5d124f-ba2d-469d-a9a0-3d33ee184b0d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NTE2OTUsIm5iZiI6MTcwNzc1MTM5NSwicGF0aCI6Ii8xNTA3MzEyOC8zMDQxMzQ1MzYtOGM1ZDEyNGYtYmEyZC00NjlkLWE5YTAtM2QzM2VlMTg0YjBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDE1MjMxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxMmY3OTczNWFhMTA1NDU1ZjMyMDliMjRkY2Q3MmFhZmY2OTA2YmI1OTJlMzU5ZTA0Yjg4NTE1Y2IxMWFmYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xY_I2TJY1tF-4RB4A7PUwijmQFUsyY4M5JDVwy4Tji4">

There's is work in the making that will enhance this select control with more info to help users. But will be worked in a separated PR.

Antonio-RiveroMartnez commented:
> 1. Since "Filters (custom)" control works only if Range For Comparison is "Custom", can we make the visibility of that control dynamic, so that it's actually only visible when it's useful? I believe we use such pattern in some other charts.
> 2. Can we remove the collapsible and maybe change the name of the control to sth like "Custom comparison range"?
> 
> <img alt="image" width="327" src="https://private-user-images.githubusercontent.com/15073128/304135471-ce13795d-8095-4598-ad94-55ae06a8a177.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc3NTE3NjIsIm5iZiI6MTcwNzc1MTQ2MiwicGF0aCI6Ii8xNTA3MzEyOC8zMDQxMzU0NzEtY2UxMzc5NWQtODA5NS00NTk4LWFkOTQtNTVhZTA2YThhMTc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjEyVDE1MjQyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkNTRmYzMyOWQ1Y2U1N2E2NmU1MmIyODFjYzFhNjdjNTMxYTFiYjY2MGE1M2VkMzFmZWE4ZTc2YmEzZjE2ODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ph7i3UXHBB3OckSND5t7MS1Ouz2g9EOAAboC7gy8Ym4">

Yes, this PR ha snot been updated with latest master but a PR got merged addressing this already https://github.com/apache/superset/pull/27054

kgabryje commented:
That's a TOTAL nitpick but... the name "huge" sounds a bit overdramatic since the text isn't really that big 😆 

<img width="1197" alt="image" src="https://github.com/apache/superset/assets/15073128/4aebc8ac-5230-4285-8273-ddc449e954b0">

kgabryje reviewed (approved):
LGTM, the bugs I listed can be worked on in the follow up PRs

Antonio-RiveroMartnez committed:
BigNumber with Time Comparison:

- Extract comparison period calculation to an util file and make use of it when building the query. This is to support following work that will make use of that function and to prepare us when adding proper testing.
